File: #2. 2019 카카오 개발자 겨울 인턴십/김민정/problem4_징검다리건너기.java
// 풀이 시간: 80분
// 시간 복잡도: O(N log N) (N : stones배열 길이)
// 코드 설명: stones에 나오는 숫자들을 treeset에 넣어서 중복제거, 정렬하고 이진탐색을 사용하여 가능한 가장 작은 값을 찾음
// 취약한 부분: 이분탐색.. 생각하는데 오래 걸림 효율성 검사 통과하는데 오래 걸렸어요


import java.util.*;

class Solution {
    public int solution(int[] stones, int k) {
        
        TreeSet<Integer> ts = new TreeSet<>();
        for (int s : stones) ts.add(s);
        List<Integer> list = new ArrayList<>(ts);
        
        int left = 0;
        int right = ts.size()-1;
        while (left < right) {
            int mid = (left + right) / 2;
            
            if (check(list.get(mid), k, stones)) {
                right = mid;
            } else left = mid+1;
        }
        
        return list.get(left);
    }
    
    public boolean check(int n, int k, int[] stones) {
        int cnt = 0;
        for (int s : stones) {
            if (s <= n) {
                cnt++;
                if (cnt == k) return true;
            } else cnt = 0;
        }
        
        return false;
    }
}

File: #2. 2019 카카오 개발자 겨울 인턴십/장효빈/problem4_징검다리_건너기.py
# 풀이 시간: 60분
# 시간 복잡도: log(2억) * 20만
# 코드 설명: 이분탐색으로 매 mid 마다 가능한 지 확인하여 최대 인원을 찾기
# 취약한 부분: 정해는 20~30분 안에 찾았으나 효율성에서 통과하지 못함. 왜 그런지 모르겠음. 민정이 코드보고 수정.

def solution(stones, k):
    def is_possible(mid):
        cnt = 0
        for stone in stones:
            if stone < mid:
                cnt += 1
                if cnt == k:
                    return False
            else:
                cnt = 0
            
        return True
    
    l = 1
    r = 200000000
    answer = 1
    while l <= r:
        mid = (l + r) // 2
        if is_possible(mid):
            answer = mid
            l = mid + 1
        else:
            r = mid - 1
    
    return answer